File: SRC/zlar1v.f
      SUBROUTINE ZLAR1V( N, B1, BN, LAMBDA, D, L, LD, LLD, PIVMIN, GAPTOL, Z, WANTNC, NEGCNT, ZTZ, MINGMA, R, ISUPPZ, NRMINV, RESID, RQCORR, WORK )
*
*  -- LAPACK auxiliary routine --
*  -- LAPACK is a software package provided by Univ. of Tennessee,    --
*  -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--
*
*     .. Scalar Arguments ..
      bool               WANTNC;
      int       B1, BN, N, NEGCNT, R;
      double             GAPTOL, LAMBDA, MINGMA, NRMINV, PIVMIN, RESID, RQCORR, ZTZ;
*     ..
*     .. Array Arguments ..
      int                ISUPPZ( * );
      double             D( * ), L( * ), LD( * ), LLD( * ), WORK( * );
      COMPLEX*16       Z( * )
*     ..
*
*  =====================================================================
*
*     .. Parameters ..
      double             ZERO, ONE;
      PARAMETER          ( ZERO = 0.0D0, ONE = 1.0D0 )
      COMPLEX*16         CONE
      PARAMETER          ( CONE = ( 1.0D0, 0.0D0 ) )

*     ..
*     .. Local Scalars ..
      bool               SAWNAN1, SAWNAN2;
      int                I, INDLPL, INDP, INDS, INDUMN, NEG1, NEG2, R1, R2;
      double             DMINUS, DPLUS, EPS, S, TMP;
*     ..
*     .. External Functions ..
      bool    DISNAN;
      double             DLAMCH;
      EXTERNAL           DISNAN, DLAMCH
*     ..
*     .. Intrinsic Functions ..
      // INTRINSIC ABS, DBLE
*     ..
*     .. Executable Statements ..
*
      EPS = DLAMCH( 'Precision' )


      IF( R.EQ.0 ) THEN
         R1 = B1
         R2 = BN
      ELSE
         R1 = R
         R2 = R
      END IF

*     Storage for LPLUS
      INDLPL = 0
*     Storage for UMINUS
      INDUMN = N
      INDS = 2*N + 1
      INDP = 3*N + 1

      IF( B1.EQ.1 ) THEN
         WORK( INDS ) = ZERO
      ELSE
         WORK( INDS+B1-1 ) = LLD( B1-1 )
      END IF

*
*     Compute the stationary transform (using the differential form)
*     until the index R2.
*
      SAWNAN1 = .FALSE.
      NEG1 = 0
      S = WORK( INDS+B1-1 ) - LAMBDA
      DO 50 I = B1, R1 - 1
         DPLUS = D( I ) + S
         WORK( INDLPL+I ) = LD( I ) / DPLUS
         IF(DPLUS.LT.ZERO) NEG1 = NEG1 + 1
         WORK( INDS+I ) = S*WORK( INDLPL+I )*L( I )
         S = WORK( INDS+I ) - LAMBDA
 50   CONTINUE
      SAWNAN1 = DISNAN( S )
      IF( SAWNAN1 ) GOTO 60
      DO 51 I = R1, R2 - 1
         DPLUS = D( I ) + S
         WORK( INDLPL+I ) = LD( I ) / DPLUS
         WORK( INDS+I ) = S*WORK( INDLPL+I )*L( I )
         S = WORK( INDS+I ) - LAMBDA
 51   CONTINUE
      SAWNAN1 = DISNAN( S )
*
 60   CONTINUE
      IF( SAWNAN1 ) THEN
*        Runs a slower version of the above loop if a NaN is detected
         NEG1 = 0
         S = WORK( INDS+B1-1 ) - LAMBDA
         DO 70 I = B1, R1 - 1
            DPLUS = D( I ) + S
            IF(ABS(DPLUS).LT.PIVMIN) DPLUS = -PIVMIN
            WORK( INDLPL+I ) = LD( I ) / DPLUS
            IF(DPLUS.LT.ZERO) NEG1 = NEG1 + 1
            WORK( INDS+I ) = S*WORK( INDLPL+I )*L( I )
            IF( WORK( INDLPL+I ).EQ.ZERO ) WORK( INDS+I ) = LLD( I )
            S = WORK( INDS+I ) - LAMBDA
 70      CONTINUE
         DO 71 I = R1, R2 - 1
            DPLUS = D( I ) + S
            IF(ABS(DPLUS).LT.PIVMIN) DPLUS = -PIVMIN
            WORK( INDLPL+I ) = LD( I ) / DPLUS
            WORK( INDS+I ) = S*WORK( INDLPL+I )*L( I )
            IF( WORK( INDLPL+I ).EQ.ZERO ) WORK( INDS+I ) = LLD( I )
            S = WORK( INDS+I ) - LAMBDA
 71      CONTINUE
      END IF
*
*     Compute the progressive transform (using the differential form)
*     until the index R1
*
      SAWNAN2 = .FALSE.
      NEG2 = 0
      WORK( INDP+BN-1 ) = D( BN ) - LAMBDA
      DO 80 I = BN - 1, R1, -1
         DMINUS = LLD( I ) + WORK( INDP+I )
         TMP = D( I ) / DMINUS
         IF(DMINUS.LT.ZERO) NEG2 = NEG2 + 1
         WORK( INDUMN+I ) = L( I )*TMP
         WORK( INDP+I-1 ) = WORK( INDP+I )*TMP - LAMBDA
 80   CONTINUE
      TMP = WORK( INDP+R1-1 )
      SAWNAN2 = DISNAN( TMP )

      IF( SAWNAN2 ) THEN
*        Runs a slower version of the above loop if a NaN is detected
         NEG2 = 0
         DO 100 I = BN-1, R1, -1
            DMINUS = LLD( I ) + WORK( INDP+I )
            IF(ABS(DMINUS).LT.PIVMIN) DMINUS = -PIVMIN
            TMP = D( I ) / DMINUS
            IF(DMINUS.LT.ZERO) NEG2 = NEG2 + 1
            WORK( INDUMN+I ) = L( I )*TMP
            WORK( INDP+I-1 ) = WORK( INDP+I )*TMP - LAMBDA
            IF( TMP.EQ.ZERO ) WORK( INDP+I-1 ) = D( I ) - LAMBDA
 100     CONTINUE
      END IF
*
*     Find the index (from R1 to R2) of the largest (in magnitude)
*     diagonal element of the inverse
*
      MINGMA = WORK( INDS+R1-1 ) + WORK( INDP+R1-1 )
      IF( MINGMA.LT.ZERO ) NEG1 = NEG1 + 1
      IF( WANTNC ) THEN
         NEGCNT = NEG1 + NEG2
      ELSE
         NEGCNT = -1
      ENDIF
      IF( ABS(MINGMA).EQ.ZERO ) MINGMA = EPS*WORK( INDS+R1-1 )
      R = R1
      DO 110 I = R1, R2 - 1
         TMP = WORK( INDS+I ) + WORK( INDP+I )
         IF( TMP.EQ.ZERO ) TMP = EPS*WORK( INDS+I )
         IF( ABS( TMP ).LE.ABS( MINGMA ) ) THEN
            MINGMA = TMP
            R = I + 1
         END IF
 110  CONTINUE
*
*     Compute the FP vector: solve N^T v = e_r
*
      ISUPPZ( 1 ) = B1
      ISUPPZ( 2 ) = BN
      Z( R ) = CONE
      ZTZ = ONE
*
*     Compute the FP vector upwards from R
*
      IF( .NOT.SAWNAN1 .AND. .NOT.SAWNAN2 ) THEN
         DO 210 I = R-1, B1, -1
            Z( I ) = -( WORK( INDLPL+I )*Z( I+1 ) )
            IF( (ABS(Z(I))+ABS(Z(I+1)))* ABS(LD(I)).LT.GAPTOL ) THEN
               Z( I ) = ZERO
               ISUPPZ( 1 ) = I + 1
               GOTO 220
            ENDIF
            ZTZ = ZTZ + DBLE( Z( I )*Z( I ) )
 210     CONTINUE
 220     CONTINUE
      ELSE
*        Run slower loop if NaN occurred.
         DO 230 I = R - 1, B1, -1
            IF( Z( I+1 ).EQ.ZERO ) THEN
               Z( I ) = -( LD( I+1 ) / LD( I ) )*Z( I+2 )
            ELSE
               Z( I ) = -( WORK( INDLPL+I )*Z( I+1 ) )
            END IF
            IF( (ABS(Z(I))+ABS(Z(I+1)))* ABS(LD(I)).LT.GAPTOL ) THEN
               Z( I ) = ZERO
               ISUPPZ( 1 ) = I + 1
               GO TO 240
            END IF
            ZTZ = ZTZ + DBLE( Z( I )*Z( I ) )
 230     CONTINUE
 240     CONTINUE
      ENDIF

*     Compute the FP vector downwards from R in blocks of size BLKSIZ
      IF( .NOT.SAWNAN1 .AND. .NOT.SAWNAN2 ) THEN
         DO 250 I = R, BN-1
            Z( I+1 ) = -( WORK( INDUMN+I )*Z( I ) )
            IF( (ABS(Z(I))+ABS(Z(I+1)))* ABS(LD(I)).LT.GAPTOL ) THEN
               Z( I+1 ) = ZERO
               ISUPPZ( 2 ) = I
               GO TO 260
            END IF
            ZTZ = ZTZ + DBLE( Z( I+1 )*Z( I+1 ) )
 250     CONTINUE
 260     CONTINUE
      ELSE
*        Run slower loop if NaN occurred.
         DO 270 I = R, BN - 1
            IF( Z( I ).EQ.ZERO ) THEN
               Z( I+1 ) = -( LD( I-1 ) / LD( I ) )*Z( I-1 )
            ELSE
               Z( I+1 ) = -( WORK( INDUMN+I )*Z( I ) )
            END IF
            IF( (ABS(Z(I))+ABS(Z(I+1)))* ABS(LD(I)).LT.GAPTOL ) THEN
               Z( I+1 ) = ZERO
               ISUPPZ( 2 ) = I
               GO TO 280
            END IF
            ZTZ = ZTZ + DBLE( Z( I+1 )*Z( I+1 ) )
 270     CONTINUE
 280     CONTINUE
      END IF
*
*     Compute quantities for convergence test
*
      TMP = ONE / ZTZ
      NRMINV = SQRT( TMP )
      RESID = ABS( MINGMA )*NRMINV
      RQCORR = MINGMA*TMP
*
*
      RETURN
*
*     End of ZLAR1V
*
      END

File: lib/src/zlar1v.dart
import 'dart:math';

import 'package:lapack/src/box.dart';
import 'package:lapack/src/complex.dart';
import 'package:lapack/src/disnan.dart';
import 'package:lapack/src/install/dlamch.dart';
import 'package:lapack/src/matrix.dart';

void zlar1v(
  final int N,
  final int B1,
  final int BN,
  final double LAMBDA,
  final Array<double> D_,
  final Array<double> L_,
  final Array<double> LD_,
  final Array<double> LLD_,
  final double PIVMIN,
  final double GAPTOL,
  final Array<Complex> Z_,
  final bool WANTNC,
  final Box<int> NEGCNT,
  final Box<double> ZTZ,
  final Box<double> MINGMA,
  final Box<int> R,
  final Array<int> ISUPPZ_,
  final Box<double> NRMINV,
  final Box<double> RESID,
  final Box<double> RQCORR,
  final Array<double> WORK_,
) {
// -- LAPACK auxiliary routine --
// -- LAPACK is a software package provided by Univ. of Tennessee,    --
// -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--
  final Z = Z_.having();
  final D = D_.having();
  final L = L_.having();
  final LD = LD_.having();
  final LLD = LLD_.having();
  final ISUPPZ = ISUPPZ_.having();
  final WORK = WORK_.having();
  const ZERO = 0.0, ONE = 1.0;
  bool SAWNAN1, SAWNAN2;
  int I, INDLPL, INDP, INDS, INDUMN, NEG1, NEG2, R1, R2;
  double DMINUS, DPLUS, EPS, S, TMP;

  EPS = dlamch('Precision');

  if (R.value == 0) {
    R1 = B1;
    R2 = BN;
  } else {
    R1 = R.value;
    R2 = R.value;
  }

  // Storage for LPLUS
  INDLPL = 0;
  // Storage for UMINUS
  INDUMN = N;
  INDS = 2 * N + 1;
  INDP = 3 * N + 1;

  if (B1 == 1) {
    WORK[INDS] = ZERO;
  } else {
    WORK[INDS + B1 - 1] = LLD[B1 - 1];
  }

  // Compute the stationary transform (using the differential form)
  // until the index R2.

  SAWNAN1 = false;
  NEG1 = 0;
  S = WORK[INDS + B1 - 1] - LAMBDA;
  for (I = B1; I <= R1 - 1; I++) {
    DPLUS = D[I] + S;
    WORK[INDLPL + I] = LD[I] / DPLUS;
    if (DPLUS < ZERO) NEG1++;
    WORK[INDS + I] = S * WORK[INDLPL + I] * L[I];
    S = WORK[INDS + I] - LAMBDA;
  }
  SAWNAN1 = disnan(S);
  if (!SAWNAN1) {
    for (I = R1; I <= R2 - 1; I++) {
      DPLUS = D[I] + S;
      WORK[INDLPL + I] = LD[I] / DPLUS;
      WORK[INDS + I] = S * WORK[INDLPL + I] * L[I];
      S = WORK[INDS + I] - LAMBDA;
    }
    SAWNAN1 = disnan(S);
  }

  if (SAWNAN1) {
    // Runs a slower version of the above loop if a NaN is detected
    NEG1 = 0;
    S = WORK[INDS + B1 - 1] - LAMBDA;
    for (I = B1; I <= R1 - 1; I++) {
      DPLUS = D[I] + S;
      if (DPLUS.abs() < PIVMIN) DPLUS = -PIVMIN;
      WORK[INDLPL + I] = LD[I] / DPLUS;
      if (DPLUS < ZERO) NEG1++;
      WORK[INDS + I] = S * WORK[INDLPL + I] * L[I];
      if (WORK[INDLPL + I] == ZERO) WORK[INDS + I] = LLD[I];
      S = WORK[INDS + I] - LAMBDA;
    }
    for (I = R1; I <= R2 - 1; I++) {
      DPLUS = D[I] + S;
      if (DPLUS.abs() < PIVMIN) DPLUS = -PIVMIN;
      WORK[INDLPL + I] = LD[I] / DPLUS;
      WORK[INDS + I] = S * WORK[INDLPL + I] * L[I];
      if (WORK[INDLPL + I] == ZERO) WORK[INDS + I] = LLD[I];
      S = WORK[INDS + I] - LAMBDA;
    }
  }

  // Compute the progressive transform (using the differential form)
  // until the index R1

  SAWNAN2 = false;
  NEG2 = 0;
  WORK[INDP + BN - 1] = D[BN] - LAMBDA;
  for (I = BN - 1; I >= R1; I--) {
    DMINUS = LLD[I] + WORK[INDP + I];
    TMP = D[I] / DMINUS;
    if (DMINUS < ZERO) NEG2++;
    WORK[INDUMN + I] = L[I] * TMP;
    WORK[INDP + I - 1] = WORK[INDP + I] * TMP - LAMBDA;
  }
  TMP = WORK[INDP + R1 - 1];
  SAWNAN2 = disnan(TMP);

  if (SAWNAN2) {
    // Runs a slower version of the above loop if a NaN is detected
    NEG2 = 0;
    for (I = BN - 1; I >= R1; I--) {
      DMINUS = LLD[I] + WORK[INDP + I];
      if (DMINUS.abs() < PIVMIN) DMINUS = -PIVMIN;
      TMP = D[I] / DMINUS;
      if (DMINUS < ZERO) NEG2++;
      WORK[INDUMN + I] = L[I] * TMP;
      WORK[INDP + I - 1] = WORK[INDP + I] * TMP - LAMBDA;
      if (TMP == ZERO) WORK[INDP + I - 1] = D[I] - LAMBDA;
    }
  }

  // Find the index (from R1 to R2) of the largest (in magnitude)
  // diagonal element of the inverse

  MINGMA.value = WORK[INDS + R1 - 1] + WORK[INDP + R1 - 1];
  if (MINGMA.value < ZERO) NEG1++;
  if (WANTNC) {
    NEGCNT.value = NEG1 + NEG2;
  } else {
    NEGCNT.value = -1;
  }
  if ((MINGMA.value).abs() == ZERO) MINGMA.value = EPS * WORK[INDS + R1 - 1];
  R.value = R1;
  for (I = R1; I <= R2 - 1; I++) {
    TMP = WORK[INDS + I] + WORK[INDP + I];
    if (TMP == ZERO) TMP = EPS * WORK[INDS + I];
    if (TMP.abs() <= (MINGMA.value).abs()) {
      MINGMA.value = TMP;
      R.value = I + 1;
    }
  }

  // Compute the FP vector: solve N^T v = e_r

  ISUPPZ[1] = B1;
  ISUPPZ[2] = BN;
  Z[R.value] = Complex.one;
  ZTZ.value = ONE;

  // Compute the FP vector upwards from R.value

  if (!SAWNAN1 && !SAWNAN2) {
    for (I = R.value - 1; I >= B1; I--) {
      Z[I] = -(WORK[INDLPL + I].toComplex() * Z[I + 1]);
      if ((Z[I].abs() + Z[I + 1].abs()) * LD[I].abs() < GAPTOL) {
        Z[I] = Complex.zero;
        ISUPPZ[1] = I + 1;
        break;
      }
      ZTZ.value += (Z[I] * Z[I]).toDouble();
    }
  } else {
    // Run slower loop if NaN occurred.
    for (I = R.value - 1; I >= B1; I--) {
      if (Z[I + 1] == Complex.zero) {
        Z[I] = -(LD[I + 1] / LD[I]).toComplex() * Z[I + 2];
      } else {
        Z[I] = -(WORK[INDLPL + I].toComplex() * Z[I + 1]);
      }
      if ((Z[I].abs() + Z[I + 1].abs()) * LD[I].abs() < GAPTOL) {
        Z[I] = Complex.zero;
        ISUPPZ[1] = I + 1;
        break;
      }
      ZTZ.value += (Z[I] * Z[I]).toDouble();
    }
  }

  // Compute the FP vector downwards from R.value in blocks of size BLKSIZ
  if (!SAWNAN1 && !SAWNAN2) {
    for (I = R.value; I <= BN - 1; I++) {
      Z[I + 1] = -(WORK[INDUMN + I].toComplex() * Z[I]);
      if ((Z[I].abs() + Z[I + 1].abs()) * LD[I].abs() < GAPTOL) {
        Z[I + 1] = Complex.zero;
        ISUPPZ[2] = I;
        break;
      }
      ZTZ.value += (Z[I + 1] * Z[I + 1]).toDouble();
    }
  } else {
    // Run slower loop if NaN occurred.
    for (I = R.value; I <= BN - 1; I++) {
      if (Z[I] == Complex.zero) {
        Z[I + 1] = -(LD[I - 1] / LD[I]).toComplex() * Z[I - 1];
      } else {
        Z[I + 1] = -(WORK[INDUMN + I].toComplex() * Z[I]);
      }
      if ((Z[I].abs() + Z[I + 1].abs()) * LD[I].abs() < GAPTOL) {
        Z[I + 1] = Complex.zero;
        ISUPPZ[2] = I;
        break;
      }
      ZTZ.value += (Z[I + 1] * Z[I + 1]).toDouble();
    }
  }

  // Compute quantities for convergence test

  TMP = ONE / ZTZ.value;
  NRMINV.value = sqrt(TMP);
  RESID.value = (MINGMA.value).abs() * NRMINV.value;
  RQCORR.value = MINGMA.value * TMP;
}
